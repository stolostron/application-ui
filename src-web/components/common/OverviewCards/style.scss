/*******************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2019. All Rights Reserved.
 * Copyright (c) 2020 Red Hat, Inc.
 * US Government Users Restricted Rights - Use, duplication or disclosure
 * restricted by GSA ADP Schedule Contract with IBM Corp.
 *******************************************************************************/

.overview-cards-container {
  // Details accordion
  .overview-cards-details-section {
    background-color: white;
    box-shadow: 0 1px 2px 0 rgba(0,0,0,0.1);
  }

  // Accordion header
  .bx--accordion__heading {
    height: 57px;
    padding: 10px;
    border-bottom: 1px solid rgba(0,0,0,0.1);
  }

  // Accordion title
  .bx--accordion__title {
    font-size: 16px;
    font-weight: 600;
    color: #151515;
  }

  // Accordion arrow
  .bx--accordion__arrow {
    fill: #151515 !important;
    position: absolute;
    right: 7%;
  }

  // App Overview details
  .bx--accordion__item--active>.bx--accordion__content {
    height: 170px;
  }

  // Divide left and right sides
  .bx--accordion__content {
    display: grid;
    grid-template-columns: 50% 50%;
    padding: 0;
  }

  .details-col {
    display: grid;
    grid-template-rows: 40px 40px 40px;
    padding: calc(10px + 1rem);
  }

  #left-col {
    grid-column-start: 1;
    border-right: 1px solid rgba(0,0,0,0.1);
  }

  #right-col {
    grid-column-start: 2;
  }

  // The 3 rows of data for each side
  .details-item {
    padding: 10px 0;
    font-size: 14px;
  }

  .details-item-title {
    float: left;
    font-weight: 600;

    &.left-item {
      width: 20%;
    }
    &.right-item {
      width: 32%;
    }
  }

  .loading-skeleton-text {
    display: inline-block;
  }

  // Cluster resource status icons
  .status-icon-container {
    display: inline-block;
    width: 40px;
    height: 24px;
    border-radius: 12px;
    padding: 6px 8px;
    margin-right: 8px;
  }

  .green-status {
    background-color: #F3FAF2;
  }

  .yellow-status {
    background-color: #FDF7E7;
  }

  .red-status {
    background-color: #FAEAE8;
  }

  .orange-status {
    background-color: #F0F0F0;
  }

  .status-icon {
    width: 12px;
    height: 12px;
  }

  .status-count {
    display: inline-block;
    font-size: 12px;
    margin-left: 4px;
    vertical-align: top;
  }

  // Search resource link
  .details-item-link {
    color: #0066CC;
    text-decoration: none;
  }

  .details-item-link-icon {
    margin: 0 10px;
    height: 10px;
  }
}

.resource-cards-info,
.overview-cards-info {
  display: flex;
  padding: 0 5%;
  height: 100%;
  justify-content: space-between;
  align-items: center;

  .card-margin {
    margin: 0 7%;
  }

  .clickable {
    &:hover {
      cursor: pointer;
    }
  }

  .loading-skeleton-text {
    &-header {
      margin-left: 20%;
      height: 20px;
    }

    &-lrg {
      margin-left: 10%;
      height: 10px;
    }

    &-med {
      margin-left: 20%;
      height: 10px;
    }

    &-sm {
      margin-left: 30%;
      margin-top: 5px;
      height: 10px;
    }
  }

  .card-count {
    line-height: 44px;
    font-size: 34px;
    font-weight: 600;
    margin-bottom: 5px;
    text-align: center;
    color: #5a6872;

    &.alert {
      color: #e0182d;
    }
  }

  .card-type {
    letter-spacing: 1.03px;
    text-align: center;
    font-size: 12px;
    font-weight: 600;
    line-height: 20px;
    color: #5a6872;
    text-transform: uppercase;
  }

  .card-text {
    text-align: center;
    font-size: 12px;
    font-style: italic;
    line-height: 20px;
    color: #4a4a4a;
  }

  .card-subtext {
    text-align: center;
    font-size: 12px;
    line-height: 20px;
    color: #4a4a4a;
  }

  .column-divider {
    height: 100%;
    border-left: 1px solid rgba(0, 0, 0, 0.1);
  }

  .row-divider {
    width: 100%;
    border-top: 1px solid rgba(0, 0, 0, 0.1);
    margin-top: 5px;
    padding-top: 5px;
  }
}
