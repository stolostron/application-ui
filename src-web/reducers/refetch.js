/*******************************************************************************
 * Copyright (c) 2020 Red Hat, Inc.
 *******************************************************************************/

import { REFETCH_INTERVAL_UPDATE } from '../actions'

export const refetch = (state, action) => {
  switch (action.type) {
  case REFETCH_INTERVAL_UPDATE:
    return action.data
  default:
    // get the poll interval from RefreshTimeSelect component
    return Object.assign({}, state, {})
  }
}
