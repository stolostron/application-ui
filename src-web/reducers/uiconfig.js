/** *****************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2017, 2019. All Rights Reserved.
 *
 * US Government Users Restricted Rights - Use, duplication or disclosure
 * restricted by GSA ADP Schedule Contract with IBM Corp.
 *******************************************************************************/

import { UICONFIG_RECEIVE_SUCCESS } from '../actions'

export const uiconfig = (state = null, action) => {
  switch (action.type) {
  case UICONFIG_RECEIVE_SUCCESS:
    return action.data
  default:
    return state
  }
}
