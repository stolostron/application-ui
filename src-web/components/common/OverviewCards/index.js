/*******************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2019. All Rights Reserved.
 * Copyright (c) 2020 Red Hat, Inc.
 *
 * US Government Users Restricted Rights - Use, duplication or disclosure
 * restricted by GSA ADP Schedule Contract with IBM Corp.
 *******************************************************************************/

import React from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import {
  Accordion,
  AccordionItem,
  Button,
  Icon,
  SkeletonText
} from 'carbon-components-react'
import { RESOURCE_TYPES } from '../../../../lib/shared/constants'
import { fetchResource } from '../../../actions/common'
import resources from '../../../../lib/shared/resources'
import { fetchTopology } from '../../../actions/topology'
import msgs from '../../../../nls/platform.properties'
import config from '../../../../lib/shared/config'
import {
  getSearchLinkForOneApplication,
  getAppOverviewCardsData
} from '../ResourceOverview/utils'
import {
  startPolling,
  stopPolling,
  handleRefreshPropertiesChanged,
  handleVisibilityChanged
} from '../../../shared/utils/refetch'

/* eslint-disable react/prop-types */

resources(() => {
  require('./style.scss')
})

const mapDispatchToProps = (dispatch, ownProps) => {
  const { selectedAppNS, selectedAppName } = ownProps
  return {
    fetchApplication: () =>
      dispatch(
        fetchResource(
          RESOURCE_TYPES.QUERY_APPLICATIONS,
          selectedAppNS,
          selectedAppName
        )
      ),
    fetchAppTopology: (fetchChannel, reloading) => {
      const fetchFilters = {
        application: { selectedAppName, selectedAppNS, channel: fetchChannel }
      }
      dispatch(
        fetchTopology({ filter: { ...fetchFilters } }, fetchFilters, reloading)
      )
    }
  }
}

const mapStateToProps = state => {
  const { QueryApplicationList, topology } = state
  return {
    QueryApplicationList,
    topology
  }
}

class OverviewCards extends React.Component {
  static propTypes = {
    fetchAppTopology: PropTypes.func
  };

  constructor(props) {
    super(props)
    this.state = {
      nodeStatuses: { green: 0, yellow: 0, red: 0, orange: 0 },
      showSubCards: false
    }
  }

  componentDidMount() {
    const { fetchApplication, fetchAppTopology } = this.props
    const activeChannel = '__ALL__/__ALL__//__ALL__/__ALL__'
    fetchAppTopology(activeChannel, true)
    fetchApplication()

    document.addEventListener('visibilitychange', this.onVisibilityChange)
    startPolling(this, setInterval)
  }

  componentWillUnmount() {
    stopPolling(this.state, clearInterval)
    document.removeEventListener('visibilitychange', this.onVisibilityChange)
  }

  onVisibilityChange = () => {
    handleVisibilityChanged(this, clearInterval, setInterval)
  };

  componentDidUpdate(prevProps) {
    handleRefreshPropertiesChanged(prevProps, this, clearInterval, setInterval)
  }

  render() {
    const {
      QueryApplicationList,
      topology,
      selectedAppName,
      selectedAppNS
    } = this.props
    const { nodeStatuses, showSubCards } = this.state
    const { locale } = this.context

    let getUrl = window.location.href
    getUrl = getUrl.substring(0, getUrl.indexOf('/multicloud/applications/'))

    const targetLink = getSearchLinkForOneApplication({
      name: encodeURIComponent(selectedAppName),
      namespace: encodeURIComponent(selectedAppNS)
    })

    const appOverviewCardsData = getAppOverviewCardsData(
      QueryApplicationList,
      topology,
      selectedAppName,
      selectedAppNS,
      nodeStatuses,
      targetLink
    )

    let clusterString = ''
    if (appOverviewCardsData.localClusterDeploy) {
      if (appOverviewCardsData.remoteClusterCount > 0) {
        const tempString =
          appOverviewCardsData.remoteClusterCount +
          ' Remote, 1 Local deployment'
        clusterString = tempString
      } else {
        clusterString = 'Local deployment'
      }
    } else {
      clusterString = appOverviewCardsData.remoteClusterCount + ' Remote'
    }

    const disableBtn =
      appOverviewCardsData.subsList && appOverviewCardsData.subsList.length > 0
        ? false
        : true

    return (
      <div className="overview-cards-container">
        <Accordion>
          <AccordionItem
            open={true}
            title={msgs.get('dashboard.card.overview.cards.title', locale)}
            className="overview-cards-details-section"
          >
            <div className="details-col" id="left-col">
              <div className="details-item">
                <div className="details-item-title left-item">
                  {msgs.get('dashboard.card.overview.cards.name', locale)}
                </div>
                <div className="details-item-content">
                  {appOverviewCardsData.appName}
                </div>
              </div>

              <div className="details-item">
                <div className="details-item-title left-item">
                  {msgs.get('dashboard.card.overview.cards.namespace', locale)}
                </div>
                <div className="details-item-content">
                  {appOverviewCardsData.appNamespace}
                </div>
              </div>

              <div className="details-item">
                <div className="details-item-title left-item">
                  {msgs.get('dashboard.card.overview.cards.created', locale)}
                </div>
                <div className="details-item-content">
                  {this.renderData(
                    appOverviewCardsData.creationTimestamp,
                    appOverviewCardsData.creationTimestamp,
                    '30%'
                  )}
                </div>
              </div>
            </div>

            <div className="details-col" id="right-col">
              <div className="details-item">
                <div className="details-item-title right-item">
                  {msgs.get('dashboard.card.overview.cards.clusters', locale)}
                </div>
                <div className="details-item-content">
                  {this.renderData(
                    appOverviewCardsData.remoteClusterCount !== -1 ||
                    appOverviewCardsData.localClusterDeploy
                      ? 0
                      : -1,
                    clusterString,
                    '30%'
                  )}
                </div>
              </div>

              <div className="details-item">
                <div className="details-item-title right-item">
                  {msgs.get(
                    'dashboard.card.overview.cards.cluster.resource.status',
                    locale
                  )}
                </div>
                <div className="details-item-content">
                  {this.renderData(
                    appOverviewCardsData.nodeStatuses,
                    this.createStatusIcons(appOverviewCardsData.nodeStatuses),
                    '30%'
                  )}
                </div>
              </div>

              <div className="details-item">
                <a
                  className="details-item-link"
                  id="app-search-link"
                  href={getUrl + appOverviewCardsData.targetLink}
                  target="_blank"
                >
                  <div>
                    {msgs.get(
                      'dashboard.card.overview.cards.search.resource',
                      locale
                    )}
                    <Icon
                      name="icon--arrow--right"
                      fill="#0066CC"
                      description=""
                      className="details-item-link-icon"
                    />
                  </div>
                </a>
              </div>
            </div>
          </AccordionItem>
        </Accordion>

        <div className="overview-cards-subs-section">
          {showSubCards
            ? this.createSubsCards(appOverviewCardsData.subsList, locale)
            : ''}
          <Button
            className="toggle-subs-btn"
            disabled={disableBtn}
            onClick={() => this.toggleSubsBtn(showSubCards)}
          >
            {this.renderData(
              appOverviewCardsData.subsList,
              (showSubCards
                ? msgs.get(
                  'dashboard.card.overview.cards.subs.btn.hide',
                  locale
                )
                : msgs.get(
                  'dashboard.card.overview.cards.subs.btn.show',
                  locale
                )) + ` (${appOverviewCardsData.subsList.length})`,
              '70%'
            )}
          </Button>
        </div>
      </div>
    )
  }

  renderData = (checkData, showData, width) => {
    return checkData !== -1 ? (
      showData
    ) : (
      <SkeletonText width={width} className="loading-skeleton-text" />
    )
  };

  createStatusIcons = nodeStatuses => {
    return (
      <React.Fragment>
        <div className="status-icon-container green-status">
          <img
            className="status-icon"
            alt="node-status-success"
            src={`${config.contextPath}/graphics/nodeStatusSuccess.svg`}
          />
          <div className="status-count">{nodeStatuses.green}</div>
        </div>
        <div className="status-icon-container yellow-status">
          <img
            className="status-icon"
            alt="node-status-warning"
            src={`${config.contextPath}/graphics/nodeStatusWarning.svg`}
          />
          <div className="status-count">{nodeStatuses.yellow}</div>
        </div>
        <div className="status-icon-container red-status">
          <img
            className="status-icon"
            alt="node-status-failure"
            src={`${config.contextPath}/graphics/nodeStatusFailure.svg`}
          />
          <div className="status-count">{nodeStatuses.red}</div>
        </div>
        <div className="status-icon-container orange-status">
          <img
            className="status-icon"
            alt="node-status-pending"
            src={`${config.contextPath}/graphics/nodeStatusPending.svg`}
          />
          <div className="status-count">{nodeStatuses.orange}</div>
        </div>
      </React.Fragment>
    )
  };

  createSubsCards = (subsList, locale) => {
    return subsList.map(sub => {
      if (sub.name) {
        return (
          <React.Fragment key={sub.id}>
            <div className="sub-card-container">
              <div className="sub-card-column">
                <Icon
                  name="icon--filter--glyph"
                  fill="#5c5c5c"
                  description=""
                  className="subs-icon"
                />
                <div className="sub-card-content">
                  <div className="sub-card-title">
                    {msgs.get(
                      'dashboard.card.overview.cards.subs.label',
                      locale
                    )}
                  </div>
                  <span>{sub.name}</span>
                </div>
              </div>

              <div className="sub-card-column">
                <Icon
                  name="icon--folder"
                  fill="#5c5c5c"
                  description=""
                  className="subs-icon"
                />
                <div className="sub-card-content">
                  <div className="sub-card-title">
                    {msgs.get(
                      'dashboard.card.overview.cards.repoResource.label',
                      locale
                    )}
                  </div>
                  <div className="sub-card-status-icon" id="resource-type-icon">
                    <a
                      className="resource-type-link"
                      href={sub.resourcePath}
                      target="_blank"
                    >
                      {sub.resourceType}
                      <Icon
                        name="icon--launch"
                        description=""
                        className="resource-type-icon"
                      />
                    </a>
                  </div>
                </div>
              </div>

              <div className="sub-card-column">
                <Icon
                  name="icon--terminal"
                  fill="#5c5c5c"
                  description=""
                  className="subs-icon"
                />
                <div className="sub-card-content">
                  <div className="sub-card-title">
                    {msgs.get(
                      'dashboard.card.overview.cards.timeWindow.label',
                      locale
                    )}
                  </div>
                  {sub.timeWindowType === 'default' ? (
                    <a
                      className="set-time-window-link"
                      href={
                        window.location.href +
                        (window.location.href.slice(-1) === '/'
                          ? 'yaml'
                          : '/yaml')
                      }
                      target="_blank"
                    >
                      {msgs.get(
                        'dashboard.card.overview.cards.timeWindow.set.label',
                        locale
                      )}
                    </a>
                  ) : (
                    <div
                      className="sub-card-status-icon"
                      id={sub.timeWindowType + '-type-icon'}
                    >
                      {sub.timeWindowType}
                    </div>
                  )}
                </div>
              </div>
            </div>
          </React.Fragment>
        )
      }
      return ''
    })
  };

  toggleSubsBtn = showSubCards => {
    this.setState({ showSubCards: !showSubCards })
    this.forceUpdate()
  };
}

OverviewCards.propTypes = {}

export default connect(mapStateToProps, mapDispatchToProps)(OverviewCards)
