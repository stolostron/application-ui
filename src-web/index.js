/*******************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2017, 2019. All Rights Reserved.
 * Copyright (c) 2020 Red Hat, Inc.
 *
 * US Government Users Restricted Rights - Use, duplication or disclosure
 * restricted by GSA ADP Schedule Contract with IBM Corp.
 *******************************************************************************/
import React from 'react'
import { hydrate } from 'react-dom'
import { Provider } from 'react-redux'
import thunkMiddleware from 'redux-thunk'
import { createLogger } from 'redux-logger'
import { createStore, combineReducers, applyMiddleware, compose } from 'redux'
import { Router } from 'react-router-dom'
import { ApolloProvider } from 'react-apollo'
import App from './containers/App'
import ScrollToTop from './components/common/ScrollToTop'
import * as reducers from './reducers'
import config from '../lib/shared/config'
import apolloClient from '../lib/client/apollo-client'
import history from 'history'

const composeEnhancers = window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose

window.SHARED_HISTORY = window.SHARED_HISTORY
  ? window.SHARED_HISTORY
  : history.createBrowserHistory()

const loggerMiddleware = createLogger()
// Grab the state from a global variable injected into the server-generated HTML
const preloadedState = window.__PRELOADED_STATE__

const middleware = [thunkMiddleware] // lets us dispatch() functions
if (
  !window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ &&
  config['featureFlags:reduxLogger']
) {
  middleware.push(loggerMiddleware) // middleware that logs actions
}

// Allow the passed state to be garbage-collected
delete window.__PRELOADED_STATE__

// Create Redux store with initial state
const store = createStore(
  combineReducers(reducers),
  preloadedState,
  composeEnhancers(applyMiddleware(...middleware))
)

hydrate(
  <ApolloProvider client={apolloClient.getClient()}>
    <Provider store={store}>
      <Router history={window.SHARED_HISTORY}>
        <ScrollToTop>
          <App />
        </ScrollToTop>
      </Router>
    </Provider>
  </ApolloProvider>,
  document.getElementById('page')
)
