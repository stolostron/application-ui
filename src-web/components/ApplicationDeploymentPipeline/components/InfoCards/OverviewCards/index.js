/*******************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2019. All Rights Reserved.
 * Copyright (c) 2020 Red Hat, Inc.
 *
 * US Government Users Restricted Rights - Use, duplication or disclosure
 * restricted by GSA ADP Schedule Contract with IBM Corp.
 *******************************************************************************/

import R from 'ramda'
import React from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import { SkeletonText } from 'carbon-components-react'
import resources from '../../../../../../lib/shared/resources'
import { RESOURCE_TYPES } from '../../../../../../lib/shared/constants'
import { fetchResources } from '../../../../../actions/common'
import { bindActionCreators } from 'redux'
import * as Actions from '../../../../../actions'
import msgs from '../../../../../../nls/platform.properties'
import {
  getNumIncidents,
  getSubscriptionDataOnHub,
  getSubscriptionDataOnManagedClustersSingle,
  getPodData,
  getIncidentsData,
  concatDataForTextKey,
  concatDataForSubTextKey
} from '../utils'
import { getSearchLinkForOneApplication } from '../../../../common/ResourceOverview/utils'
import config from '../../../../../../lib/shared/config'

/* eslint-disable react/prop-types */

resources(() => {
  require('../style.scss')
})

const mapDispatchToProps = dispatch => {
  return {
    fetchApplications: () =>
      dispatch(fetchResources(RESOURCE_TYPES.QUERY_APPLICATIONS)),
    actions: bindActionCreators(Actions, dispatch)
  }
}

const mapStateToProps = state => {
  const {
    QueryApplicationList,
    CEMIncidentList,
    secondaryHeader,
    AppOverview
  } = state
  const isSingleApplicationView =
    R.pathOr([], ['breadcrumbItems'])(secondaryHeader).length === 2
  const enableCEM = AppOverview.showCEMAction
  return {
    QueryApplicationList,
    CEMIncidentList,
    isSingleApplicationView,
    enableCEM
  }
}

const getOverviewCardsData = (
  QueryApplicationList,
  CEMIncidentList,
  isSingleApplicationView,
  enableCEM,
  applicationName,
  applicationNamespace,
  targetLinkForSubscriptions,
  targetLinkForClusters,
  targetLinkForPods,
  locale
) => {
  // All functions will return -1 if data or data.items is undefined... this will be used for skeleton text load bar
  const subscriptionDataOnHub = getSubscriptionDataOnHub(
    QueryApplicationList,
    isSingleApplicationView,
    applicationName,
    applicationNamespace
  )
  var subscriptionDataOnManagedClusters = getSubscriptionDataOnManagedClustersSingle(
    QueryApplicationList,
    applicationName,
    applicationNamespace
  )
  const podData = getPodData(
    QueryApplicationList,
    applicationName,
    applicationNamespace
  )
  const incidents = getNumIncidents(CEMIncidentList)
  const incidentsData = getIncidentsData(CEMIncidentList)

  const result = [
    {
      msgKey:
        subscriptionDataOnHub.total === 1
          ? msgs.get('dashboard.card.deployment.subscription', locale)
          : msgs.get('dashboard.card.deployment.subscriptions', locale),
      count: subscriptionDataOnHub.total,
      targetLink:
        subscriptionDataOnHub.total === 0 ? '' : targetLinkForSubscriptions,
      textKey: msgs.get('dashboard.card.deployment.subscriptions.text', locale),
      subtextKeyFirst: concatDataForSubTextKey(
        subscriptionDataOnHub.total,
        subscriptionDataOnHub.total,
        subscriptionDataOnHub.failed,
        msgs.get('dashboard.card.deployment.failed.lowercase', locale)
      ),
      subtextKeySecond: concatDataForSubTextKey(
        subscriptionDataOnHub.total,
        subscriptionDataOnHub.noStatus,
        subscriptionDataOnHub.noStatus,
        msgs.get('dashboard.card.deployment.noStatus', locale)
      )
    },
    {
      msgKey:
        subscriptionDataOnManagedClusters.clusters === 1
          ? msgs.get('dashboard.card.deployment.managedCluster', locale)
          : msgs.get('dashboard.card.deployment.managedClusters', locale),
      count: subscriptionDataOnManagedClusters.clusters,
      targetLink:
        subscriptionDataOnManagedClusters.clusters === 0
          ? ''
          : targetLinkForClusters,
      textKey: concatDataForTextKey(
        subscriptionDataOnManagedClusters.clusters,
        subscriptionDataOnManagedClusters.total,
        msgs.get('dashboard.card.deployment.totalSubscription', locale),
        msgs.get('dashboard.card.deployment.totalSubscriptions', locale)
      ),
      subtextKeyFirst: concatDataForSubTextKey(
        subscriptionDataOnManagedClusters.clusters,
        subscriptionDataOnManagedClusters.clusters,
        subscriptionDataOnManagedClusters.failed,
        msgs.get('dashboard.card.deployment.failed.lowercase', locale)
      ),
      subtextKeySecond: concatDataForSubTextKey(
        subscriptionDataOnManagedClusters.clusters,
        subscriptionDataOnManagedClusters.noStatus,
        subscriptionDataOnManagedClusters.noStatus,
        msgs.get('dashboard.card.deployment.noStatus', locale)
      )
    },
    {
      msgKey:
        podData.total === 1
          ? msgs.get('dashboard.card.deployment.pod', locale)
          : msgs.get('dashboard.card.deployment.pods', locale),
      count: podData.total,
      targetLink: podData.total === 0 ? '' : targetLinkForPods,
      subtextKeyFirst: concatDataForSubTextKey(
        podData.total,
        podData.total,
        podData.running,
        msgs.get('dashboard.card.deployment.running', locale)
      ),
      subtextKeySecond: concatDataForSubTextKey(
        podData.total,
        podData.total,
        podData.failed,
        msgs.get('dashboard.card.deployment.failed.lowercase', locale)
      )
    }
  ]

  if (enableCEM) {
    result.push({
      msgKey:
        incidents === 1
          ? msgs.get('dashboard.card.deployment.incident', locale)
          : msgs.get('dashboard.card.deployment.incidents', locale),
      count: incidents,
      alert: incidents > 0 ? true : false,
      targetTab: incidents === 0 ? null : 2,
      subtextKeyFirst: concatDataForSubTextKey(
        incidents,
        incidents,
        incidentsData.priority1,
        msgs.get('dashboard.card.deployment.incidents.priority1', locale)
      ),
      subtextKeySecond: concatDataForSubTextKey(
        incidents,
        incidents,
        incidentsData.priority2,
        msgs.get('dashboard.card.deployment.incidents.priority2', locale)
      )
    })
  }
  return result
}

class OverviewCards extends React.Component {
  componentDidMount() {
    const { fetchApplications } = this.props

    fetchApplications()

    this.startPolling()
    document.addEventListener('visibilitychange', this.onVisibilityChange)
  }

  componentWillUnmount() {
    this.stopPolling()
    document.removeEventListener('visibilitychange', this.onVisibilityChange)
  }

  startPolling() {
    if (parseInt(config['featureFlags:liveUpdates'], 10) === 2) {
      var intervalId = setInterval(
        this.reload.bind(this),
        config['featureFlags:liveUpdatesPollInterval']
      )
      this.setState({ intervalId: intervalId })
    }
  }

  stopPolling() {
    if (this.state && this.state.intervalId) {
      clearInterval(this.state.intervalId)
    }
  }

  onVisibilityChange = () => {
    if (document.visibilityState === 'visible') {
      this.startPolling()
    } else {
      this.stopPolling()
    }
  };

  reload() {
    const { fetchApplications } = this.props
    fetchApplications()
  }

  render() {
    const {
      QueryApplicationList,
      CEMIncidentList,
      isSingleApplicationView,
      enableCEM,
      actions,
      selectedAppName,
      selectedAppNS
    } = this.props
    const { locale } = this.context
    const singleAppStyle = isSingleApplicationView ? ' single-app' : ''
    const applicationName = selectedAppName
    const applicationNamespace = selectedAppNS

    const targetLinkForSubscriptions = getSearchLinkForOneApplication({
      name: encodeURIComponent(applicationName),
      showRelated: 'subscription'
    })
    const targetLinkForClusters = getSearchLinkForOneApplication({
      name: encodeURIComponent(applicationName),
      showRelated: 'cluster'
    })
    const targetLinkForPods = getSearchLinkForOneApplication({
      name: encodeURIComponent(applicationName),
      showRelated: 'pod'
    })

    const overviewCardsData = getOverviewCardsData(
      QueryApplicationList,
      CEMIncidentList,
      isSingleApplicationView,
      enableCEM,
      applicationName,
      applicationNamespace,
      targetLinkForSubscriptions,
      targetLinkForClusters,
      targetLinkForPods,
      locale
    )

    return (
      <div className={'overview-cards-info' + singleAppStyle}>
        <InfoCards overviewCardsData={overviewCardsData} actions={actions} />
      </div>
    )
  }
}

const showCardData = (cardCount, width, className) => {
  return cardCount !== -1 ? (
    cardCount
  ) : (
    <SkeletonText width={width} className={className} />
  )
}

const InfoCards = ({ overviewCardsData, actions }) => {
  return (
    <React.Fragment>
      {Object.keys(overviewCardsData).map(key => {
        const card = overviewCardsData[key]
        const id = `${key}_overviewCardsData`
        var cemStatus = 'card-cem-disabled'
        if (overviewCardsData.length === 4) {
          cemStatus = 'card-cem-enabled'
        }
        const handleClick = (e, resource) => {
          if (resource.targetTab != null) {
            actions.setSelectedAppTab(resource.targetTab)
          } else if (resource.targetLink) {
            window.open(resource.targetLink, '_blank')
          }
        }
        const handleKeyPress = (e, resource) => {
          if (e.key === 'Enter') {
            handleClick(e, resource)
          }
        }

        return (
          <React.Fragment key={key}>
            <div
              id={id}
              key={card}
              className={
                card.targetLink || card.targetTab
                  ? `single-card clickable ${cemStatus}`
                  : `single-card ${cemStatus}`
              }
              role="button"
              tabIndex="0"
              onClick={e => handleClick(e, card)}
              onKeyPress={e => handleKeyPress(e, card)}
            >
              <div className={card.alert ? 'card-count alert' : 'card-count'}>
                {showCardData(
                  card.count,
                  '60%',
                  'loading-skeleton-text-header'
                )}
              </div>
              <div className="card-type">{card.msgKey}</div>
              <div className="card-text">
                {showCardData(card.textKey, '80%', 'loading-skeleton-text-lrg')}
              </div>
              {(card.subtextKeyFirst || card.subtextKeySecond) && (
                <div className="row-divider" />
              )}
              {card.subtextKeyFirst && (
                <div className="card-subtext">
                  {showCardData(
                    card.subtextKeyFirst,
                    '40%',
                    'loading-skeleton-text-sm'
                  )}
                </div>
              )}
              {card.subtextKeySecond && (
                <div className="card-subtext">
                  {showCardData(
                    card.subtextKeySecond,
                    '60%',
                    'loading-skeleton-text-med'
                  )}
                </div>
              )}
            </div>
            {key < Object.keys(overviewCardsData).length - 1 && (
              <div className="column-divider" />
            )}
          </React.Fragment>
        )
      })}
    </React.Fragment>
  )
}

InfoCards.propTypes = {
  actions: PropTypes.object,
  overviewCardsData: PropTypes.array
}

OverviewCards.propTypes = {}

export default connect(mapStateToProps, mapDispatchToProps)(OverviewCards)
