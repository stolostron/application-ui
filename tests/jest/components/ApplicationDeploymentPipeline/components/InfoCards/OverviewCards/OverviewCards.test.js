/*******************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2019. All Rights Reserved.
 *
 * US Government Users Restricted Rights - Use, duplication or disclosure
 * restricted by GSA ADP Schedule Contract with IBM Corp.
 *******************************************************************************/

const React = require("../../../../../../../node_modules/react");

import OverviewCards from "../../../../../../../src-web/components/ApplicationDeploymentPipeline/components/InfoCards/OverviewCards";

import renderer from "react-test-renderer";
import * as reducers from "../../../../../../../src-web/reducers";

import { createStore, combineReducers, applyMiddleware, compose } from "redux";
import thunkMiddleware from "redux-thunk";
import { Provider } from "react-redux";

const preloadedState = window.__PRELOADED_STATE__;
const composeEnhancers = window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose;
const middleware = [thunkMiddleware];

const store = createStore(
  combineReducers(reducers),
  preloadedState,
  composeEnhancers(applyMiddleware(...middleware))
);

describe("OverviewCards", () => {
  it("OverviewCards renders correctly with data.", () => {
    const tree = renderer
      .create(
        <Provider store={store}>
          <OverviewCards selectedAppName="app1" selectedAppNS="default" />
        </Provider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
