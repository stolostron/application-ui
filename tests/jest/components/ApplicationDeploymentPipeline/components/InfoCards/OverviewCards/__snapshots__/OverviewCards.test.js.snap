// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OverviewCards OverviewCards renders correctly with data. 1`] = `
<div
  className="overview-cards-info"
>
  <div
    className="single-card"
    onClick={[Function]}
    onKeyPress={[Function]}
    role="button"
    tabIndex="0"
  >
    <div
      className="card-count"
    >
      0
    </div>
    <div
      className="card-type"
    >
      Subscriptions
    </div>
    <div
      className="card-text"
    >
      On hub cluster
    </div>
    
    
    
  </div>
  <div
    className="column-divider"
  />
  <div
    className="single-card"
    onClick={[Function]}
    onKeyPress={[Function]}
    role="button"
    tabIndex="0"
  >
    <div
      className="card-count"
    >
      0
    </div>
    <div
      className="card-type"
    >
      Managed clusters
    </div>
    <div
      className="card-text"
    >
      0 Total subscriptions
    </div>
    
    
    
  </div>
  <div
    className="column-divider"
  />
  <div
    className="single-card"
    onClick={[Function]}
    onKeyPress={[Function]}
    role="button"
    tabIndex="0"
  >
    <div
      className="card-count"
    >
      0
    </div>
    <div
      className="card-type"
    >
      Pods
    </div>
    <div
      className="card-text"
    />
    
    
    
  </div>
</div>
`;
